File: rs/dagrun/src/profile_enrich.rs
//! Per-step measurement enrichment from cgroup and ambient-load observations.

// Per-step measurement ENRICHMENT: fold a step's cgroup + ambient deltas into the rich
// profile-row columns the parallel-speedup model reads back (Rust port of
// `py/dagrun/profile_enrich.py`).
//
// The column NAMES mirror the originating `scripts/validate_perflog.py` `STEP_PROFILE_COLUMNS`
// (`effective_cores`, `quota_utilization_pct`, `throttled_s`, `external_cpu_s`/`external_cores`,
// `co_tenants_*`, `ambient_bucket`, `load*`, host/step PSI), so a later schema unification is a
// RENAME, not a redesign. Everything is best-effort and captured UNDER cgroup boxing; an
// unavailable measurement simply leaves its column out of the returned row and the writer records
// it blank (No Silent Failure).
//
// These are WRITER columns whose exact numeric values legitimately differ per host/run, so they
// are NOT byte-compared across the Python and Rust builds (only the CSV schema is). The READER
// that consumes them (`estimates::load_step_speedups`) IS cross-checked byte-identical.

use std::collections::BTreeMap;

use crate::ambient::{ambient_bucket, attribute_external_cores, AmbientSnapshot, PsiReading};
use crate::perflog::{effective_cpu_quota, nproc};

const USEC: f64 = 1_000_000.0;

/// Return the effective core budget of the current cgroup or process affinity.
///
/// Walks the current cgroup's ancestor chain for the tightest finite CPU quota, floors it to a
/// positive whole-core budget, and takes the tighter of that quota and process affinity. Flooring
/// is deliberate: a per-step width ceiling must never promise more CPU than the binding
/// bandwidth cap, and integer arithmetic avoids language-specific rounding at half-core boundaries.
pub fn container_core_budget() -> i64 {
    let affinity = nproc().max(1);
    core_budget_from_quota(&effective_cpu_quota(), affinity)
}

fn core_budget_from_quota(quota: &str, affinity: i64) -> i64 {
    let affinity = affinity.max(1);
    let mut parts = quota.split('_');
    let quota_cores = match (parts.next(), parts.next(), parts.next()) {
        (Some(q), Some(p), None) => match (q.parse::<i64>(), p.parse::<i64>()) {
            (Ok(q), Ok(p)) if q > 0 && p > 0 => Some((q / p).max(1)),
            _ => None,
        },
        _ => None,
    };
    quota_cores.map_or(affinity, |quota| quota.min(affinity))
}

/// Resolve an optional step width to a concrete positive-environment budget.
///
/// An explicit width is returned unchanged; otherwise the current container core budget is used.
pub fn resolve_effective_inner_jobs(inner_jobs: Option<i64>) -> i64 {
    inner_jobs.unwrap_or_else(container_core_budget)
}

fn psi_columns(
    row: &mut BTreeMap<String, String>,
    prefix: &str,
    start: Option<&PsiReading>,
    end: Option<&PsiReading>,
) {
    if let Some(s) = start {
        row.insert(
            format!("{prefix}_psi_avg10_start"),
            format!("{:.2}", s.avg10),
        );
        row.insert(
            format!("{prefix}_psi_avg60_start"),
            format!("{:.2}", s.avg60),
        );
    }
    if let Some(e) = end {
        row.insert(format!("{prefix}_psi_avg10_end"), format!("{:.2}", e.avg10));
        row.insert(format!("{prefix}_psi_avg60_end"), format!("{:.2}", e.avg60));
    }
}

/// Build rich profile columns from one step's cgroup and ambient measurements.
///
/// Only columns whose inputs are available are returned; the profile writer leaves the remaining
/// fields in [`crate::perflog::STEP_PROFILE_COLUMNS`] empty.
#[allow(clippy::too_many_arguments)]
pub fn step_enrichment_columns(
    elapsed_s: f64,
    inner_jobs: Option<i64>,
    cpu_stats: Option<&BTreeMap<String, i64>>,
    ambient_start: Option<&AmbientSnapshot>,
    ambient_end: Option<&AmbientSnapshot>,
    step_pressure_start: Option<&PsiReading>,
    step_pressure_end: Option<&PsiReading>,
) -> BTreeMap<String, String> {
    let mut row: BTreeMap<String, String> = BTreeMap::new();
    if let Some(stats) = cpu_stats {
        if elapsed_s > 0.0 {
            let usage_usec = *stats.get("usage_usec").unwrap_or(&0);
            let effective_cores = usage_usec as f64 / (elapsed_s * USEC);
            row.insert("effective_cores".into(), format!("{effective_cores:.4}"));
            row.insert(
                "user_s".into(),
                format!("{:.3}", *stats.get("user_usec").unwrap_or(&0) as f64 / USEC),
            );
            row.insert(
                "sys_s".into(),
                format!(
                    "{:.3}",
                    *stats.get("system_usec").unwrap_or(&0) as f64 / USEC
                ),
            );
            row.insert(
                "throttled_s".into(),
                format!(
                    "{:.3}",
                    *stats.get("throttled_usec").unwrap_or(&0) as f64 / USEC
                ),
            );
            if let Some(j) = inner_jobs {
                if j != 0 {
                    row.insert(
                        "quota_utilization_pct".into(),
                        format!("{:.2}", effective_cores / j as f64 * 100.0),
                    );
                }
            }
            if let (Some(a0), Some(a1)) = (ambient_start, ambient_end) {
                let external_cores = attribute_external_cores(
                    a0.busy_jiffies,
                    a1.busy_jiffies,
                    usage_usec,
                    elapsed_s,
                );
                row.insert(
                    "external_cpu_s".into(),
                    format!("{:.3}", external_cores * elapsed_s),
                );
                row.insert("external_cores".into(), format!("{external_cores:.3}"));
                row.insert(
                    "ambient_bucket".into(),
                    ambient_bucket(external_cores, a1).value().to_string(),
                );
            }
        }
    }
    if let (Some(a0), Some(a1)) = (ambient_start, ambient_end) {
        row.insert("co_tenants_start".into(), a0.co_tenants.to_string());
        row.insert("co_tenants_end".into(), a1.co_tenants.to_string());
        row.insert("load1_start".into(), format!("{:.3}", a0.load1));
        row.insert("load1_end".into(), format!("{:.3}", a1.load1));
        row.insert("load5_start".into(), format!("{:.3}", a0.load5));
        row.insert("load5_end".into(), format!("{:.3}", a1.load5));
        psi_columns(
            &mut row,
            "host_cpu",
            a0.cpu_psi.as_ref(),
            a1.cpu_psi.as_ref(),
        );
        psi_columns(
            &mut row,
            "host_memory",
            a0.memory_psi.as_ref(),
            a1.memory_psi.as_ref(),
        );
        psi_columns(&mut row, "host_io", a0.io_psi.as_ref(), a1.io_psi.as_ref());
    }
    psi_columns(&mut row, "step_cpu", step_pressure_start, step_pressure_end);
    row
}

#[cfg(test)]
mod tests {
    use super::core_budget_from_quota;

    #[test]
    fn fractional_cpu_quota_is_conservatively_floored() {
        assert_eq!(core_budget_from_quota("50000_100000", 64), 1);
        assert_eq!(core_budget_from_quota("150000_100000", 64), 1);
        assert_eq!(core_budget_from_quota("250000_100000", 64), 2);
        assert_eq!(core_budget_from_quota("350000_100000", 2), 2);
        assert_eq!(core_budget_from_quota("max", 7), 7);
        assert_eq!(core_budget_from_quota("unknown", 7), 7);
    }
}

File: py/dagrun/profile_enrich.py
"""Enrich per-step profile rows with cgroup and ambient-load measurements.

The pure enrichment functions derive effective parallelism, throttling, pressure,
co-tenant activity, and peak-memory columns without fabricating unavailable values.
"""

from __future__ import annotations

from collections.abc import Mapping
from dagrun import perflog
from dagrun.ambient import (
    AmbientSnapshot,
    PsiReading,
    ambient_bucket,
    attribute_external_cores,
)

__all__ = [
    "step_enrichment_columns",
    "resolve_effective_inner_jobs",
    "container_core_budget",
]

_USEC = 1_000_000.0

#: cgroup-v2 unified CPU quota file for the whole run's container/machine.
def container_core_budget() -> int:
    """The effective core count of the cgroup/container/machine the run is boxed in.

    Walks the current cgroup's ancestor chain for the tightest finite CPU quota, floors it to a
    positive whole-core count, and takes the tighter of that quota and the process affinity
    width. This is the NUMBER an "ambient" (un-``-j``-capped) step's effective parallelism
    resolves to — never the string ``"ambient"``.
    """
    affinity = max(1, perflog.nproc())
    quota = perflog.effective_cpu_quota()
    parts = quota.split("_")
    if len(parts) != 2:
        return affinity
    try:
        quota_us, period_us = (int(part) for part in parts)
    except ValueError:
        return affinity
    if quota_us <= 0 or period_us <= 0:
        return affinity
    # A whole-job default must never overstate a fractional bandwidth grant. A positive sub-core
    # quota still gets one job so the run can make progress.
    quota_cores = max(1, quota_us // period_us)
    return min(affinity, quota_cores)


def resolve_effective_inner_jobs(inner_jobs: int | None) -> int:
    """Resolve a step's recorded ``inner_jobs`` to a NUMBER.

    An explicit ``preferred_inner_jobs`` width is used verbatim; an "ambient" step (no explicit
    ``-j``) resolves to :func:`container_core_budget` — the effective parallelism of the
    cgroup/container/machine it ran in. This replaces the old string ``"ambient"`` sentinel so the
    speedup model can group samples by a real parallelism level.
    """
    if inner_jobs is not None:
        return inner_jobs
    return container_core_budget()


def _psi_columns(
    row: dict[str, object],
    prefix: str,
    start: PsiReading | None,
    end: PsiReading | None,
) -> None:
    """Fill the four ``<prefix>_psi_avg{10,60}_{start,end}`` columns from two PSI readings.

    An absent reading leaves its columns out of ``row`` (the writer records them blank), so a
    kernel without PSI is distinguishable from genuine zero pressure."""
    if start is not None:
        row[f"{prefix}_psi_avg10_start"] = round(start.avg10, 2)
        row[f"{prefix}_psi_avg60_start"] = round(start.avg60, 2)
    if end is not None:
        row[f"{prefix}_psi_avg10_end"] = round(end.avg10, 2)
        row[f"{prefix}_psi_avg60_end"] = round(end.avg60, 2)


def step_enrichment_columns(
    *,
    elapsed_s: float,
    inner_jobs: int | None,
    cpu_stats: Mapping[str, int] | None,
    ambient_start: AmbientSnapshot | None,
    ambient_end: AmbientSnapshot | None,
    step_pressure_start: PsiReading | None,
    step_pressure_end: PsiReading | None,
) -> dict[str, object]:
    """Build the rich per-step profile columns from one step's cgroup + ambient measurements.

    Only columns whose inputs are present are returned; the caller's CSV writer fills the rest
    blank (they are part of :data:`dagrun.perflog.STEP_PROFILE_COLUMNS`). Derivations:

    * ``effective_cores`` = child CPU-seconds / wall (``cpu.stat`` ``usage_usec`` / ``elapsed_s``),
      the ACHIEVED parallelism; ``user_s`` / ``sys_s`` / ``throttled_s`` from the matching
      ``cpu.stat`` counters.
    * ``quota_utilization_pct`` = ``effective_cores`` / the applied inner ``-j`` cap (only when an
      explicit ``inner_jobs`` cap was set — an ambient step has no cap to measure against).
    * ``external_cpu_s`` / ``external_cores`` = host CPU burned by OTHER tenants during the step
      window (busy-jiffies delta minus our own cgroup usage), and ``ambient_bucket`` the
      quiet/moderate/busy verdict — so the reader can contention-discount this sample.
    * ``co_tenants_*`` / ``load*`` / host+step PSI = the ambient-load context around the window.
    """
    row: dict[str, object] = {}
    usage_usec = 0
    have_cpu = cpu_stats is not None and elapsed_s > 0.0
    if have_cpu and cpu_stats is not None:
        usage_usec = cpu_stats.get("usage_usec", 0)
        effective_cores = usage_usec / (elapsed_s * _USEC)
        row["effective_cores"] = round(effective_cores, 4)
        row["user_s"] = round(cpu_stats.get("user_usec", 0) / _USEC, 3)
        row["sys_s"] = round(cpu_stats.get("system_usec", 0) / _USEC, 3)
        row["throttled_s"] = round(cpu_stats.get("throttled_usec", 0) / _USEC, 3)
        if inner_jobs:  # quota cores == the applied inner cpu cap (blank for an ambient step)
            row["quota_utilization_pct"] = round(effective_cores / inner_jobs * 100.0, 2)
        if ambient_start is not None and ambient_end is not None:
            external_cores = attribute_external_cores(
                busy_jiffies_start=ambient_start.busy_jiffies,
                busy_jiffies_end=ambient_end.busy_jiffies,
                own_cpu_usec=usage_usec,
                elapsed_s=elapsed_s,
            )
            # external_cpu_s == external_cores * elapsed_s (both derive from the same clamped delta).
            row["external_cpu_s"] = round(external_cores * elapsed_s, 3)
            row["external_cores"] = round(external_cores, 3)
            row["ambient_bucket"] = ambient_bucket(external_cores, ambient_end)
    if ambient_start is not None and ambient_end is not None:
        row["co_tenants_start"] = ambient_start.co_tenants
        row["co_tenants_end"] = ambient_end.co_tenants
        row["load1_start"] = round(ambient_start.load1, 3)
        row["load1_end"] = round(ambient_end.load1, 3)
        row["load5_start"] = round(ambient_start.load5, 3)
        row["load5_end"] = round(ambient_end.load5, 3)
        _psi_columns(row, "host_cpu", ambient_start.cpu_psi, ambient_end.cpu_psi)
        _psi_columns(row, "host_memory", ambient_start.memory_psi, ambient_end.memory_psi)
        _psi_columns(row, "host_io", ambient_start.io_psi, ambient_end.io_psi)
    _psi_columns(row, "step_cpu", step_pressure_start, step_pressure_end)
    return row
